### PR TITLE
Alert streams: backend

### PR DESCRIPTION
**Pull request by dmitryduev**
In this PR:

- Data model and API adjustments
- API handlers to deal with alert streams
- Tests

### CODE REVIEW

File: skyportal/models.py [hunk 1]
```diff
@@ -66,9 +66,12 @@ def process_result_value(self, value, dialect):
 class Group(Base):
     name = sa.Column(sa.String, unique=True, nullable=False)
 
-    streams = relationship('Stream', secondary='stream_groups',
+    streams = relationship('Stream', secondary='group_streams',
```
Comment (dmitryduev) [inline skyportal/models.py:69]: Changed the order since I think in terms of adding stream access to groups rather than adding groups to streams.

**Comment by dannygoldstein** [comment]
@dmitryduev Merge conflicts

**Comment by dmitryduev** [comment]
All nice and green again, @dannygoldstein!

File: data/db_demo.yaml [hunk 2]
```diff
@@ -12,6 +12,23 @@ user:
...
+streams:
+  - name: ZTF Public
+    altdata:
+      collection: ZTF_alerts
+      selector: [1]
+    =id: ztf_public
+  - name: ZTF Partnership
+    altdata:
+      collection: ZTF_alerts
+      selector: [1, 2]
```
Comment (dannygoldstein) [inline data/db_demo.yaml:24]: Why not `2`? Technically the "partnership stream" is everything with PID=2.

File: data/db_demo.yaml [hunk 3]
```diff
@@ -12,6 +12,23 @@ user:
...
+    =id: ztf_partnership
+  - name: ZTF Caltech
+    altdata:
+      collection: ZTF_alerts
+      selector: [1, 2, 3]
```
Comment (dannygoldstein) [inline data/db_demo.yaml:29]: See comment on partnership stream

File: data/db_demo.yaml [hunk 4]
```diff
@@ -12,6 +12,23 @@ user:
     roles:
       - View only
 
+streams:
+  - name: ZTF Public
+    altdata:
+      collection: ZTF_alerts
+      selector: [1]
```
Comment (dannygoldstein) [inline data/db_demo.yaml:19]: Should this be a scalar?

File: data/db_demo.yaml [hunk 5]
```diff
@@ -30,9 +47,13 @@ groups/=program_A/users:
...
     admin: false
 
+groups/=program_A/streams:
+  - stream_id: =ztf_partnership
+
 filters:
-  - query_string: sample_query_string
+  - name: Yellow transients
     group_id: =program_A
+    stream_id: =ztf_partnership
```
Comment (dannygoldstein) [inline data/db_demo.yaml:56]: `stream_ids`?

File: data/db_demo.yaml [hunk 6]
```diff
@@ -30,9 +47,13 @@ groups/=program_A/users:
   - username: fulluser@cesium-ml.org
     admin: false
 
+groups/=program_A/streams:
+  - stream_id: =ztf_partnership
+
 filters:
-  - query_string: sample_query_string
+  - name: Yellow transients
```
Comment (dannygoldstein) [inline data/db_demo.yaml:54]: Should this be versioned?

File: skyportal/handlers/api/filter.py [hunk 7]
```diff
@@ -46,6 +47,14 @@ def get(self, filter_id=None):
             f = Filter.get_if_owned_by(filter_id, self.current_user)
             if f is None:
                 return self.error("Invalid filter ID.")
+            # get stream:
+            stream = (
```
Comment (dannygoldstein) [inline skyportal/handlers/api/filter.py:51]:
```suggestion
            stream = DBSession().query(Stream).get(f.stream_id)
```

File: skyportal/handlers/api/group.py [hunk 8]
```diff
@@ -93,6 +93,27 @@ def get(self, group_id=None):
                 # Do not include User.groups to avoid circular reference
                 group['users'] = [{'id': user.id, 'username': user.username}
                                   for user in group['users']]
+                # grab streams:
+                group_streams = (
```
Comment (dannygoldstein) [inline skyportal/handlers/api/group.py:97]:
```suggestion
                streams = DBSession().query(Stream).join(GroupStream).filter(GroupStream.group_id == group_id).all()
                group['streams'] = streams
```

File: skyportal/handlers/api/group.py [hunk 9]
```diff
@@ -93,6 +93,27 @@ def get(self, group_id=None):
...
+                streams = (
+                    DBSession().query(Stream)
+                        .filter(Stream.id.in_(stream_ids))
+                        .all()
+                )
+                group['streams'] = streams
+                # grab filters:
+                filters = (
+                    DBSession().query(Filter)
+                        .filter(Filter.group_id.in_([group_id]))
```
Comment (dannygoldstein) [inline skyportal/handlers/api/group.py:112]:
```suggestion
                        .filter(Filter.group_id == group_id)
```

File: skyportal/handlers/api/group.py [hunk 10]
```diff
@@ -358,3 +379,113 @@ def delete(self, group_id, username):
...
+        stream = Stream.query.filter(Stream.id == stream_id).first()
+        if stream is None:
+            return self.error(
+                "Specified stream_id does not exist."
+            )
+        else:
+            # Add new GroupStream
+            gs = GroupStream.query.filter(
+                GroupStream.group_id == group_id
+            ).filter(
```
Comment (dannygoldstein) [inline skyportal/handlers/api/group.py:437]:
```suggestion
           ,
```

File: skyportal/handlers/api/group.py [hunk 11]
```diff
@@ -358,3 +379,113 @@ def delete(self, group_id, username):
...
+        group = Group.query.get(group_id)
+        stream = Stream.query.filter(Stream.id == int(stream_id)).first()
+        stream_id = stream.id
+        if stream is None:
+            return self.error(
+                "Specified stream_id does not exist."
+            )
+        else:
+            if group.single_user_group:
+                return self.error("Cannot delete streams from single user groups.")
```
Comment (dannygoldstein) [inline skyportal/handlers/api/group.py:485]: We should support this, for when people change institutions.

File: skyportal/handlers/api/stream.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            - in: path
+              name: filter_id
+              required: true
+              schema:
+                type: integer
+          responses:
+            200:
+              content:
+                application/json:
+                  schema: SingleFilter
```
Comment (dannygoldstein) [inline skyportal/handlers/api/stream.py:28]: Wrong return type

File: skyportal/handlers/api/stream.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+              content:
+                application/json:
+                  schema: Error
+        multiple:
+          description: Retrieve all streams
+          responses:
+            200:
+              content:
+                application/json:
+                  schema: ArrayOfFilters
```
Comment (dannygoldstein) [inline skyportal/handlers/api/stream.py:39]: Check return type

File: skyportal/handlers/api/stream.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+from baselayer.app.access import auth_or_token, permissions
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    Stream,
+)
+
+
+class StreamHandler(BaseHandler):
+    @auth_or_token
```
Comment (dannygoldstein) [inline skyportal/handlers/api/stream.py:12]: What are the use cases for exposing this endpoint to a general user?

File: skyportal/handlers/api/stream.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+            return self.error(
+                "Invalid/missing parameters: " f"{e.normalized_messages()}"
+            )
+        DBSession().add(stream)
+        DBSession().commit()
+
+        return self.success(data={"id": stream.id})
+
+    @permissions(["System admin"])
+    def patch(self, stream_id):
```
Comment (dannygoldstein) [inline skyportal/handlers/api/stream.py:107]: We have been using PUT for partial updates, although we should be using PATCH according to the HTTP standards. 

@stefanv What do you think is the appropriate path here? Use PUT for now and then open another PR changing all the PUTs to PATCHes?

File: skyportal/models.py [hunk 16]
```diff
@@ -300,9 +316,11 @@ def altitude(self, telescope, time):
 
 
 class Filter(Base):
-    query_string = sa.Column(sa.String, nullable=False, unique=False)
+    name = sa.Column(sa.String, nullable=False, unique=False)
+    stream_id = sa.Column(sa.ForeignKey("streams.id"), nullable=False)
+    stream = relationship("Stream", foreign_keys=[stream_id], back_populates="filters")
     group_id = sa.Column(sa.ForeignKey("groups.id"), nullable=False)
```
Comment (dannygoldstein) [inline skyportal/models.py:322]:
```suggestion
    group_id = sa.Column(sa.ForeignKey("groups.id"), nullable=False, ondelete='CASCADE')
```

File: skyportal/models.py [hunk 17]
```diff
@@ -96,23 +98,37 @@ class Group(Base):
...
                           passive_deletes=True)
+    users = relationship('User', secondary='stream_users',
+                         back_populates='streams',
+                         passive_deletes=True)
+    filters = relationship('Filter',
+                           back_populates='stream',
+                           passive_deletes=True)
+
 
+GroupStream = join_model('group_streams', Group, Stream)
```
Comment (dannygoldstein) [inline skyportal/models.py:116]: If a stream is deleted, will it delete all the associated users in that stream? We should ensure that doesn't happen...

File: skyportal/models.py [hunk 18]
```diff
@@ -300,9 +316,11 @@ def altitude(self, telescope, time):
 
 
 class Filter(Base):
-    query_string = sa.Column(sa.String, nullable=False, unique=False)
+    name = sa.Column(sa.String, nullable=False, unique=False)
+    stream_id = sa.Column(sa.ForeignKey("streams.id"), nullable=False)
```
Comment (dannygoldstein) [inline skyportal/models.py:320]: Shouldn't need this, as it is now specified by `group`, right? I.e., filters no longer store stream credentials.

File: skyportal/tests/api/test_groups.py [hunk 19]
```diff
@@ -120,3 +120,59 @@ def test_public_group(view_only_token):
     assert status == 200
     assert response["status"] == "success"
     int(response["data"]["id"])
+
```
Comment (dannygoldstein) [inline skyportal/tests/api/test_groups.py:123]: Can you add the following tests?

   

- [ ] Grant a stream to a user, then delete the stream, and check that the user still exists, and has access to their original streams, minus the new one

- [ ]  Grant a stream to a group, then delete the stream, and check that the group still exists, and has access to its original streams, minus the new one

- [ ] Attach a filter to a group, then delete the group, and check that the filter has also been deleted in a cascade.

File: skyportal/tests/conftest.py [hunk 20]
```diff
@@ -48,8 +54,8 @@ def public_group2():
 
 
 @pytest.fixture()
-def public_filter(public_group):
-    return FilterFactory(group=public_group)
+def public_filter(public_group, public_stream):
+    return FilterFactory(group=public_group, stream=public_stream)
```
Comment (dannygoldstein) [inline skyportal/tests/conftest.py:58]: stream should no longer be needed here

File: skyportal/tests/fixtures.py [hunk 21]
```diff
@@ -75,17 +75,28 @@ class Meta(BaseMeta):
     observed_at = datetime.datetime.now()
 
 
+class StreamFactory(factory.alchemy.SQLAlchemyModelFactory):
+    class Meta(BaseMeta):
+        model = Stream
+    name = factory.LazyFunction(lambda: str(uuid.uuid4()))
+    users = []
+    groups = []
+    filters = []
```
Comment (dannygoldstein) [inline skyportal/tests/fixtures.py:84]: filters should no longer be needed here

File: data/db_demo.yaml (hunk 2 above)
Comment (dmitryduev) [reply data/db_demo.yaml:24]: That's why I called it a "selector". This is then used on K as in `{"candidate.programid": {"$in": <stream["altdata"]["selector"]>}}`, so that a filter defined on this stream will also get executed on `programid=1` data.

File: data/db_demo.yaml (hunk 4 above)
Comment (dmitryduev) [reply data/db_demo.yaml:19]: Again, it's meant to be used with a `$in` operator. I can add an extra type check, but since it anyways lives in `altdata`, would prefer not to bother :)

File: data/db_demo.yaml (hunk 5 above)
Comment (dmitryduev) [reply data/db_demo.yaml:56]: Ah, I see what you mean.. To my taste, it's easer to define, e.g. the Partnership stream, as Public-only+Partnership-only data as opposed to Partnership-only, so that a filter is associated with a single stream (and a single group) -- this also simplifies the book-keeping on the K side a bit. After all, alerts with programid=2 come with programid=1 stuff in prv_candidates.

File: data/db_demo.yaml (hunk 6 above)
Comment (dmitryduev) [reply data/db_demo.yaml:54]: I wouldn't, but I tend to generally not care about such things :-) Are we doing that in any other situations?

Comment (dannygoldstein) [reply data/db_demo.yaml:54]: I thought we were keeping track of multiple filter versions. Is that only happening on the K side or should we do that in SP as well?

File: data/db_demo.yaml (hunk 5 above)
Comment (dannygoldstein) [reply data/db_demo.yaml:56]: This makes me think - are we going to allow streams with data from multiple instruments? E.g., a ZTF PID=2  + Gattini stream?

File: skyportal/models.py [hunk 22]
```diff
@@ -96,23 +98,37 @@ class Group(Base):
 
 class Stream(Base):
     name = sa.Column(sa.String, unique=True, nullable=False)
-    url = sa.Column(sa.String, unique=True, nullable=False)
-    username = sa.Column(sa.String)
-    password = sa.Column(sa.String)
+    altdata = sa.Column(JSONB, nullable=True,
```
Comment (dannygoldstein) [inline skyportal/models.py:101]: `altdata` doesn't seem like the right place for this. Should we also be keeping track of what instrument a stream is connected to, so we can prevent users from implementing filters that run on streams from multiple instruments / collections, if that is not supported on K?

File: data/db_demo.yaml (hunk 5 above)
Comment (dmitryduev) [reply data/db_demo.yaml:56]: In my opinion, no, it would significantly complicate the alert filtering logic (most certainly on the K side). 
This is what the groups are for (among other things): they may have access to multiple streams from different discovery engines and you would define dedicated filters on each stream.

File: data/db_demo.yaml (hunk 6 above)
Comment (dmitryduev) [reply data/db_demo.yaml:54]: Only on the K side, since that is where the action is happening -- I want to separate the concerns as much as possible and also reduce the amount of duplicated info to a bare minimum. SP really only stores `filter_id` and the corresponding `group_id`, and that is the only piece of information that is duplicated on K. That way, the AMPEL folks, for example, won't have to bother about filter versioning on the SP side either -- they would just create a filter and then post stuff to /candidates with the `filter_id`.

File: skyportal/handlers/api/filter.py (hunk 7 above)
Comment (dmitryduev) [reply skyportal/handlers/api/filter.py:51]: Changed.

File: skyportal/handlers/api/group.py (hunk 8 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py:97]: Nice, changed!

File: skyportal/handlers/api/group.py (hunk 11 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py:485]: True, removed.

File: skyportal/handlers/api/stream.py (hunk 12 above)
Comment (dmitryduev) [reply skyportal/handlers/api/stream.py:28]: Damn, fixed.

File: skyportal/handlers/api/stream.py (hunk 13 above)
Comment (dmitryduev) [reply skyportal/handlers/api/stream.py:39]: Fixed

File: skyportal/handlers/api/stream.py (hunk 14 above)
Comment (dmitryduev) [reply skyportal/handlers/api/stream.py:12]: Couldn't think of anything specific, but thought there might be, and there isn't much criminal in getting this information. But ok, changed for `@permissions(["System admin"])`.

File: skyportal/handlers/api/stream.py (hunk 15 above)
Comment (dmitryduev) [reply skyportal/handlers/api/stream.py:107]: We use PATCH for filter and user preferences updates, so there are precedents :)

File: skyportal/models.py (hunk 16 above)
Comment (dmitryduev) [reply skyportal/models.py:322]: Nice catch, fixed

File: skyportal/models.py (hunk 18 above)
Comment (dmitryduev) [reply skyportal/models.py:320]: A group may have access to multiple streams and you define a filter in the group on a stream, so I think it should be there.

File: skyportal/tests/conftest.py (hunk 20 above)
Comment (dmitryduev) [reply skyportal/tests/conftest.py:58]: See my comment above

File: skyportal/tests/fixtures.py (hunk 21 above)
Comment (dmitryduev) [reply skyportal/tests/fixtures.py:84]: See me comment above

File: skyportal/models.py (hunk 22 above)
Comment (dmitryduev) [reply skyportal/models.py:101]: I originally had two columns instead, `collection` and `selector`, but decided to move them to `altdata` (thanks to Josh for the suggestion!) to keep it general enough to be included in general SP. 

The users won't be able to do that: a filter is defined in a group for a single stream, and only the System admin can create new streams.

File: skyportal/tests/api/test_groups.py (hunk 19 above)
Comment (dmitryduev) [reply skyportal/tests/api/test_groups.py:123]: Added

File: skyportal/models.py (hunk 16 above)
Comment (dannygoldstein) [reply skyportal/models.py]: Also make sure to index all foreign keys with `index=True`, to ensure fast joins.

File: skyportal/models.py (hunk 18 above)
Comment (dannygoldstein) [reply skyportal/models.py]: In that case I think it makes more sense to enforce stream access on the filter level than the group level. Otherwise, you need to manually update each group when you remove a stream with, e.g., PID=4 (Gattini).

File: skyportal/handlers/api/group.py (hunk 9 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py]: Fixed

File: skyportal/handlers/api/group.py (hunk 10 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py]: Fixed in the commit, forgot to comment

File: skyportal/models.py (hunk 16 above)
Comment (dmitryduev) [reply skyportal/models.py]: Done.

File: skyportal/models.py (hunk 17 above)
Comment (dmitryduev) [reply skyportal/models.py:123]: This will not happen, `test_add_stream_to_single_user_group_delete_stream` in `test_groups.py` tests for that.

File: skyportal/models.py (hunk 18 above)
Comment (dmitryduev) [reply skyportal/models.py]: I don't see why you would delete a stream. But even if you do, the update will happen automatically, see `test_post_new_filter_delete_group_deletes_filter` and `test_post_new_filter_delete_stream_deletes_filter` in `test_groups.py`.

**Comment by dmitryduev** [comment]
@dannygoldstein @acrellin all tests pass locally; on the last three commits, different frontend tests fail. Any wisdom?

**Comment by dmitryduev** [comment]
@dannygoldstein @acrellin this time, `skyportal/tests/frontend/test_candidate_page.py` timed out and failed. Any suggestions?

**Comment by acrellin** [comment]
> @dannygoldstein @acrellin this time, `skyportal/tests/frontend/test_candidate_page.py` timed out and failed. Any suggestions?

@dmitryduev go ahead and mark it flaky

**Comment by dannygoldstein** [comment]
@dmitryduev see convo and resolution on #677 for current recommended
practice on flaky tests

File: data/db_demo.yaml [hunk 23]
```diff
@@ -12,6 +12,23 @@ user:
     roles:
       - View only
 
+streams:
+  - name: ZTF Public
+    altdata:
+      collection: ZTF_alerts
```
Comment (dannygoldstein) [inline data/db_demo.yaml:18]: What if I make a group that filters on streams 2 and 4 only. Will I need to make a new Stream object for 2+4 in addition to these three and grant it manually to each user and group? What if I had 2+1+4? Will I have to grant that to each user manually for them to be able to see a stream even if they have [1], and [2,4]? 

If so, I think that argues for having each stream map 1-to-1 to each PID, and granting those permissions individually to each user.

File: skyportal/handlers/api/group.py [hunk 24]
```diff
@@ -358,3 +365,110 @@ def delete(self, group_id, username):
         self.push_all(action='skyportal/REFRESH_GROUP',
                       payload={'group_id': int(group_id)})
         return self.success()
+
+
+class GroupStreamHandler(BaseHandler):
+    @permissions(['System admin'])
+    def post(self, group_id, *ignored_args):
```
Comment (dannygoldstein) [inline skyportal/handlers/api/group.py:372]: Do we need `*ignored_args`?

**Comment by dannygoldstein** [comment]
Will re-review once green

**Comment by dannygoldstein** [comment]
Also, echoing @acrellin, @dmitryduev can you please fill out a more detailed description of the PR in the original post?

File: data/db_demo.yaml [hunk 25]
```diff
@@ -12,6 +12,23 @@ user:
...
+      selector: [1]
+    =id: ztf_public
+  - name: ZTF Partnership
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply data/db_demo.yaml:21]: Sure, it's more explicit like this, added.

File: data/db_demo.yaml [hunk 26]
```diff
@@ -12,6 +12,23 @@ user:
...
+    altdata:
+      collection: ZTF_alerts
+      selector: [1, 2]
+    =id: ztf_partnership
+  - name: ZTF Caltech
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply data/db_demo.yaml:26]: Same as above, thanks

File: skyportal/handlers/api/group.py (hunk 24 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py:411]: Yes, otherwise you'd get errors like `Error in `/api/groups/8/streams`: post() takes 2 positional arguments but 3 were given` since stream id would be None see endpoint def in `app_server.py` -- `(r'/api/groups(/[0-9]+)/streams(/[0-9]+)?', GroupStreamHandler),`

File: skyportal/models.py [hunk 27]
```diff
@@ -101,23 +104,37 @@ class Group(Base):
...
+                            "`{'collection': 'ZTF_alerts', selector: [1, 2]}`")
 
-    groups = relationship('Group', secondary='stream_groups',
+    groups = relationship('Group', secondary='group_streams',
                           back_populates='streams',
                           passive_deletes=True)
+    users = relationship('User', secondary='stream_users',
+                         back_populates='streams',
+                         passive_deletes=True)
+    filters = relationship('Filter',
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply skyportal/models.py]: Thanks for bringing this up! It made me think about the ACL stuff in `api/filters.py` and I rewrote it so that now all the necessary checks are in place (I think -- needs review; updated the tests accordingly).

I strongly prefer the current scheme and believe that switching to the proposed "simplified" scheme would actually complicate things in practice, and is _a lot_ of work. If you assign streams to filters then the user who does it must have the corresponding SU permissions (remember, this is a very "privileged" task), and this will have to be done for every user filter, and we will have potentially multiple filters per program. Whereas in the current scheme, as long as you are a group admin, you can create as many filters as you like on the streams the group has access to, without asking SUs for permissions -- I think it is much more user-friendly.

I also don't quite see what needs to be "manually kept in sync". It is forbidden to change both `group_id` and `stream_id` for a filter; deleting a stream from a group deletes corresponding filters as well, see `test_post_new_filter_delete_stream_deletes_filter` in `test_groups.py`.

File: data/db_demo.yaml (hunk 23 above)
Comment (dmitryduev) [reply data/db_demo.yaml:18]: I feel that this is a valid scenario abstractly speaking, but an artificial one in practice. Concretely, what would this 2+4 mean? In practice, for F we will have ztf_public, ztf_public+partnership, ztf_public+partnership+caltech, and gattini (and in the future, LSST, and potentially other surveys). So I do not quite understand what kinds of combinations you are talking about (and in general, this all PID stuff essentially only applies to ZTF). However, the implementation here in SP is quite general and none of the `altdata` stuff is required.

Next, these streams really point at a sink on the K side (even though eventually, SP may get its own stream listeners, but it is a completely different story), this is why I "aggregate" them this way, otherwise things will be nice, abstract, and clean on the SP side, but that would add unnecessary complexity on the K side of F.

Finally, yes, when a new stream is added to the system, a SU must add it individual users first, that's the scheme we converged on. Only when/if all users of a group have access to a particular (new) stream, it can be added to the group.

File: skyportal/handlers/api/group.py (hunk 24 above)
Comment (dmitryduev) [reply skyportal/handlers/api/group.py:411]: It's similar to `GroupUserHandler` in this respect.

**Comment by dmitryduev** [comment]
:tada:
